### PR TITLE
feat(phone-number): add possibility to overwrite otp function with custom one

### PR DESCRIPTION
**Pull request by piotr-graczyk-dev**
Add support for custom OTP generation function in the phone number plugin, allowing developers to implement their own OTP generation logic while maintaining full backward compatibility. This will be very useful during testing.

## Changes Made

- ✅ Add optional `generateOTP` parameter to `PhoneNumberOptions` interface
- ✅ Update phone number plugin implementation to use custom `generateOTP` function when provided
- ✅ Maintain backward compatibility with existing default OTP generation
- ✅ Add comprehensive tests covering both regular OTP flows and password reset scenarios
- ✅ Update documentation with clear usage examples and API reference

## Problem Solved

Previously, developers were limited to the built-in OTP generation logic using numeric codes. This enhancement enables:

- **Custom OTP formats** (alphanumeric, different patterns)
- **External OTP service integration** (when using third-party generators)
- **Enhanced security requirements** (custom randomization algorithms)
- **Business-specific OTP patterns** (branded or formatted codes)

## Implementation Details

The implementation follows the existing plugin pattern:

```typescript
// New optional parameter in PhoneNumberOptions
generateOTP?: (otpLength: number) => string;

// Usage in plugin
const code = options.generateOTP
  ? options.generateOTP(opts.otpLength)
  : generateOTP(opts.otpLength);
```

## Usage Example

```typescript
import { phoneNumber } from "better-auth/plugins"

const auth = betterAuth({
  plugins: [
    phoneNumber({
      sendOTP: ({ phoneNumber, code }) => {
        // Send OTP via SMS service
      },
      generateOTP: (otpLength) => {
        // Custom OTP generation - e.g., alphanumeric
        return generateAlphanumericOTP(otpLength);
      }
    })
  ]
})
```

## Testing

- ✅ All existing tests continue to pass
- ✅ New comprehensive test suite added for custom `generateOTP` functionality
- ✅ Tests cover regular OTP sending, verification, and password reset flows
- ✅ Backward compatibility verified through existing test suite
- ✅ Edge cases tested (different OTP lengths, call counts, etc.)

## Documentation

- ✅ Updated `/docs/content/docs/plugins/phone-number.mdx` with new option
- ✅ Added clear usage examples and API documentation
- ✅ Included in the Options section with proper description

## Breaking Changes

**None** - This is a fully backward-compatible addition. All existing functionality remains unchanged.

## Related Issues

This enhancement addresses the need for flexible OTP generation while maintaining the security and reliability of the existing phone number authentication flow.
    

---

## Summary by cubic
Adds support for a custom OTP generator in the phone-number plugin so teams can plug in their own logic. Fully backward compatible.

- New Features
  - Added generateOTP?: (length) => string to PhoneNumberOptions.
  - Used the custom generator for send OTP, sign-in verification (requireVerification), and password reset.
  - Falls back to the default numeric generator when not provided.
  - Added tests for custom generators and edge cases.
  - Updated plugin docs with the new option and example.

### CODE REVIEW

**Review by karolopolko — APPROVED** [review]

File: docs/content/docs/plugins/phone-number.mdx [hunk 1]
```diff
@@ -232,6 +232,18 @@ type resetPasswordPhoneNumber = {
 
 - `otpLength`: The length of the OTP code to be generated. Default is `6`.
 - `sendOTP`: A function that sends the OTP code to the user's phone number. It takes the phone number and the OTP code as arguments.
+- `generateOTP`: A function that generates the OTP code. It takes the OTP length as an argument and returns the OTP code. If not provided, the default `generateOTP` function will be used.
+```ts
+export const auth = betterAuth({
+    plugins: [
+        phoneNumber({
+            generateOTP: (otpLength) => {
```
[the remark this answers is not in the document]
Comment (piotr-graczyk-dev) [reply docs/content/docs/plugins/phone-number.mdx:240]: I don't agree with it. People will have custom implementation of so right now I will just return custom code "123456" to return anything

File: packages/better-auth/src/plugins/phone-number/index.ts [hunk 2]
```diff
@@ -391,7 +400,9 @@ export const phoneNumber = (options?: PhoneNumberOptions) => {
 						}
 					}
 
-					const code = generateOTP(opts.otpLength);
+					const code = options.generateOTP
```
[the remark this answers is not in the document]
Comment (piotr-graczyk-dev) [reply packages/better-auth/src/plugins/phone-number/index.ts]: Done

File: packages/better-auth/src/plugins/phone-number/phone-number.test.ts [hunk 3]
```diff
@@ -437,3 +437,550 @@ describe("phone number verification requirement", async () => {
...
+
+		await client.phoneNumber.sendOtp({
+			phoneNumber: "+251911121319",
+		});
+
+		expect(capturedOtp).toHaveLength(6);
+		expect(capturedOtp).toMatch(/^\d{6}$/);
+	});
+
+	it("should use custom generateOTP for sign-in with requireVerification", async () => {
```
[the remark this answers is not in the document]
Comment (piotr-graczyk-dev) [reply packages/better-auth/src/plugins/phone-number/phone-number.test.ts]: Done nice call

**Comment by reslear** [comment]
hi @piotr-graczyk-dev by the way, look somewhere we've had related issues to close out

related: 
https://github.com/better-auth/better-auth/issues/1159
http://github.com/better-auth/better-auth/issues/2317
https://github.com/better-auth/better-auth/pull/1576

File: packages/better-auth/src/plugins/phone-number/index.ts [hunk 4]
```diff
@@ -41,6 +41,13 @@ export interface PhoneNumberOptions {
 		data: { phoneNumber: string; code: string },
 		request?: Request,
 	) => Promise<void> | void;
+	/**
+	 * Generate OTP code if not provided default generateOTP function will be used
+	 *
+	 * @param otpLength
+	 * @returns
+	 */
+	generateOTP?: (otpLength: number) => string;
```
Comment (himself65) [inline packages/better-auth/src/plugins/phone-number/index.ts:50]:
```suggestion
	generateOTP?: (otpLength: number) => string | Promise<string>;
```

Comment (himself65) [reply packages/better-auth/src/plugins/phone-number/index.ts:50]: Do we have an Awaitable type?

File: packages/better-auth/src/plugins/phone-number/phone-number.test.ts
Comment (himself65) [inline packages/better-auth/src/plugins/phone-number/phone-number.test.ts:1]: This looks like AI-generated. Actually we only need 1 test case

**Review by himself65 — CHANGES_REQUESTED** [review]

**Comment by Kylar13** [comment]
Hey! Just ran into this PR after looking for a way of adding "test numbers" to the plugin (a number that will always have the same, hardcoded OTP).

I did this in the Email OTP plugin using the `generateOTP` function and adding a guard in the `sendVerificationOTP` to not send for those same emails (unsure if there is a better way)

What is missing in order to get this over the line so I can do the same thing on phone number?

**Comment by franciscohanna92** [comment]
Hey there, how can I help to get this merged? Really needing this due to SMS costs for testing

**Comment by franciscohanna92** [comment]
@piotr-graczyk-dev could you update your PR branch with latest from `canary` branch?

**Comment by gustavovalverde** [comment]
Closing as stale. #5926 (merged) added `verifyOTP` for external verification (e.g., Twilio), which is orthogonal to this PR's `generateOTP` for custom OTP code generation. Both hooks are useful, but this branch conflicts with main and mixes `options?.generateOTP` and `opts.generateOTP` across call sites. If `generateOTP` is still wanted, file a fresh focused PR against main.
